park-calculator/src/main/scala/com/nadberezny/parkcalculator/services/FeeCalculator.scala

```diff
@@ -0,0 +1,28 @@
+package com.nadberezny.parkcalculator.services
+
+import cats.data.Kleisli
+import cats.implicits._
+import com.nadberezny.parkcalculator.db.repositories.ParkingRepository
+import com.nadberezny.parkcalculator.services.feecalculator._
+
+class FeeCalculator(implicit val parkingRepository: ParkingRepository) {
+  def calculate(vehicleId: String, currency: String): EitherCalculable = {
+    calculation.run(
+      Calculable(vehicleId = vehicleId, currency = currency)
+    )
+  }
+
+  private lazy val calculation = {
```

Comment:
Jaka są zalety użycia tego podejścia? Czy widzisz wady tego rozwiązania? Z tego co widzę, w kolejnych krokach wzbogacasz `Calculable`. Przez to ufasz, że kolejny krok się wywołał. Trochę to niebezpieczne. Np. jeśli bym w tym miejscu wyrzucił linijkę `Kleisli(getDuration) andThen` to dalej by działały kolejne kroki tylko bazowałyby na tym że `duration` było równe `0`. Nic by nie wybuchło ani przy kompilacji, ani w runtime. Jedynie testy byłyby w stanie uchronić nas przed darmowym parkowaniem dla użytkowników. Moim zdaniem trochę to stoi w opozycji do statycznego typowania.

Reply:
to prawda :|, 
wprowadziłem tutaj wzorzec, który fajnie się sprawdza, gdy ma np. ileś middlewer'ów procesujących request, wtedy przekazuje się jedną strukturę danych, a kolejność wywołania nie ma znaczenia,
wtedy wszystkie typy mogą być:
`a -> a`
(jest tak np. w Elixirowym Plug:
https://hexdocs.pm/plug/readme.html)

a tutaj z racji poprawnosci liczenia lepiej byloby miesc gwarancje na poziomie kompilacji, czyli coś w tym stylu:
```
a -> mb
b -> mc
c -> md
itd..
```

```scala
val getDuration: Calculable => Either[Error, DurationCalculable]
val getVipInfo: DurationCalculable => Either[Error, VipInfoCalculable]
val getPrice: VipInfoCalculable => Either[Error, PriceCalculable]
```